### ISSUE TITLE
Prevent loss of WIP changes by accidentally closing modals

### ISSUE THREAD

**Issue by nimasmi** [wagtail/wagtail#10899]
When a user has interacted with a modal they may have some work in progress information in the fields. Clicking outside the modal closes it, losing that work in progress information.

### Suggested behaviour

Once fields have been edited in a modal, disable the convenience behaviour, and require an explicit action. That is: do not close upon clicking outside, but still allow closing the modal by clicking the 🗙 icon or pressing the keyboard Esc key.

The underlying bootstrap-modal library has a straightforward implementation of this behaviour by setting `data-backdrop="static"` on the modal element. See [bootstrap-modal.js](https://github.com/wagtail/wagtail/blob/30ee5256e29981aa1fd1aad3c81db54b1934ea33/wagtail/admin/static_src/wagtailadmin/js/vendor/bootstrap-modal.js#L163-L164) and [Bootstrap modal docs](https://getbootstrap.com/docs/4.4/components/modal/#static-backdrop). I have confirmed this functionality works within Wagtail by modifying the modal in the browser console `this.options.backdrop = "static";` before closing.

### Also considered

Alternatively change the behaviour to add a confirmation step before closing, either upon clicking outside, or even when hitting the 🗙 icon or pressing Esc. This may be too disruptive.

### Additional context

This is a suggestion from a user, which I am relaying. I am open to discussion of whether this is a real UX improvement.

#5338 seems related.

**Comment by lb-** [comment]
An idea, with our Stimulus dialog code, we could add a change listener. If any change event bubbles from inside we know outside click should not close the modal. 

We could add a class change also, allowing us to unhide some label near the bottom of the modal.

It wouldn't be 'smart' - as in know when the user has changed fields back to their previous state.

This behaviour could be further enhanced long term with a confirmation modal/dialog before closing.

We can implement the same behaviour in the bootstrap modals, but we probably should look at having one modal implementation when possible.

There is an argument for not using outside click at all for modals we know may have forms in them. Or not using outside click at all (by default), opting in for it in small cases where the modal is known to be info only.

**Comment by nimasmi** [comment]
@lb- yes I was imagining something dumb, such as disabling the convenience close behaviour as soon as any field is modified, even if the changes are later undone.

Thanks for the context regarding Stimulus vs Bootstrap implementations. I'm aware there is refactoring work happening somewhere, but not of the details.

**Comment by lb-** [comment]
@thibaudcolas - would love some of your input here. 

Some options for a way forward.
 
1. Just change all modals to stop outside clicks
2. Stop outside clicks if we get a change event
3. Add a simple browser confirmation alert before closing (we can enhance this to a second modal in the future maybe)
4. Add a message that appears when you try to close when it looks like you've had a change, this message will advise you to manually close or something similar
5. Add some way for our models to be optionally set to static, and then change all modals that have forms to use this property.

I'm happy to do a PR as I find it super frustrating when modals close this way haha. However, I just want to get some other input before I proceed.

**Comment by thibaudcolas** [comment]
It seems to me like adding a confirmation step defeats the purpose of the "click backdrop to close" behavior, and dynamically changing this behavior based on user interaction is only going to lead to confusion.

So I’d be in favour of either leaving things as-is, or removing "click outside to close" for all modals that have an explicit backdrop and close button. @benenright what do you think?

**Comment by thibaudcolas** [comment]
We’ve discussed with @benenright, and are happy to proceed with:

> Removing "click outside to close" for all modals that have an explicit backdrop and close button

That’s specifically all full-screen modals like in use in choosers.

---

Looking at our implementation,

1. For [Bootstrap modals](https://getbootstrap.com/2.3.2/javascript.html#modals) like choosers, seems like we need to set `backdrop: 'static'`.
2. For A11yDialog / DialogController, it seems we can remove the `click->w-dialog#hide` from the [dialog.html overlay](https://github.com/wagtail/wagtail/blob/main/wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html#L15).
